pycrostates/cluster/tests/test_aahc.py

```diff
@@ -0,0 +1,1358 @@
+    caplog.set_level(logging.WARNING)
+
+    # Unfitted
+    aahCluster_ = AAHCluster(
+        n_clusters=n_clusters,
+        # ignore_polarity=True,
+        normalize_input=False,
+    )
+
+    aahCluster_.cluster_centers_  # pylint: disable=pointless-statement
+    assert "Clustering algorithm has not been fitted." in caplog.text
+    caplog.clear()
+
+    aahCluster_.info  # pylint: disable=pointless-statement
+    assert "Clustering algorithm has not been fitted." in caplog.text
+    caplog.clear()
+
+    aahCluster_.fitted_data  # pylint: disable=pointless-statement
+    assert "Clustering algorithm has not been fitted." in caplog.text
+    caplog.clear()
+
+    # Fitted
+    aahCluster_ = aah_cluster.copy()
+
+    assert aahCluster_.cluster_centers_ is not None
+    assert "Clustering algorithm has not been fitted." not in caplog.text
+    caplog.clear()
+
+    assert aahCluster_.info is not None
+    assert "Clustering algorithm has not been fitted." not in caplog.text
+    caplog.clear()
+
+    assert aahCluster_.fitted_data is not None
+    assert "Clustering algorithm has not been fitted." not in caplog.text
+    caplog.clear()
+
+    # Test fitted property
+    aahCluster_ = AAHCluster(
+        n_clusters=n_clusters,
+        # ignore_polarity=True,
+        normalize_input=False,
+    )
+    with pytest.raises(TypeError, match="'fitted' must be an instance of"):
+        aahCluster_.fitted = "101"
+    caplog.clear()
+    aahCluster_.fitted = True
+    log = "'fitted' can not be set to 'True' directly. Please use the .fit()"
+    assert log in caplog.text
+    caplog.clear()
+    aahCluster_ = aah_cluster.copy()
+    aahCluster_.fitted = True
+    log = "'fitted' can not be set to 'True' directly. The clustering"
+    assert log in caplog.text
+
+
+def test_invalid_arguments():
+    """Test invalid arguments for init and for fit."""
+    # n_clusters
+    with pytest.raises(
+        TypeError, match="'n_clusters' must be an instance of "
+    ):
+        aahCluster_ = AAHCluster(n_clusters="4")
+    with pytest.raises(ValueError, match="The number of clusters must be a"):
+        aahCluster_ = AAHCluster(n_clusters=0)
+    with pytest.raises(ValueError, match="The number of clusters must be a"):
+        aahCluster_ = AAHCluster(n_clusters=-101)
+
+    # # ignore_polarity
+    # with pytest.raises(
+    #     TypeError, match="'ignore_polarity' must be an instance of bool"
+    # ):
+    #     aahCluster_ = AAHCluster(
+    #         n_clusters=n_clusters,
+    #         ignore_polarity="asdf"
+    #     )
+    #     aahCluster_ = AAHCluster(
+    #         n_clusters=n_clusters,
+    #         ignore_polarity=None
+    #     )
+
+    # normalize_input
+    with pytest.raises(
+        TypeError, match="'normalize_input' must be an instance of bool"
+    ):
+        aahCluster_ = AAHCluster(n_clusters=n_clusters, normalize_input="asdf")
+        aahCluster_ = AAHCluster(n_clusters=n_clusters, normalize_input=None)
```

Comment:
Unless I am mistaken, the `pytest.raises` context manager will only catch the first matching exception and then skip the rest of the nested code. Take the `foo` function:

```
def foo(x):
    if x == 1:
        raise RuntimeError("test")
    else:
        return x
```

If you run:

```
import pytest

with pytest.raises(RuntimeError, match="test"):
    x = (foo(5))
    foo(1)
```

The variable `x` is defined at the end. But if you run:

```
import pytest

with pytest.raises(RuntimeError, match="test"):
    foo(1)
    x = (foo(5))
```

The variable `x` is not defined at the end. That means you can not test multiple calls that should raise the same error message in the same context manager, or only the first one will actually be tested.

Reply:
Thanks for pointing me on this issue. I updated the code accordingly.